okhttp/src/main/java/okhttp3/internal/http2/Http2Connection.java

```diff
@@ -350,60 +351,56 @@ void writeWindowUpdateLater(final int streamId, final long unacknowledgedBytesRe
         });
   }
 
-  /**
-   * Sends a ping frame to the peer. Use the returned object to await the ping's response and
-   * observe its round trip time.
-   */
-  public Ping ping() throws IOException {
-    Ping ping = new Ping();
-    int pingId;
-    synchronized (this) {
-      if (shutdown) {
-        throw new ConnectionShutdownException();
-      }
-      pingId = nextPingId;
-      nextPingId += 2;
-      if (pings == null) pings = new LinkedHashMap<>();
-      pings.put(pingId, ping);
-    }
-    writePing(false, pingId, 0x4f4b6f6b /* ASCII "OKok" */, ping);
-    return ping;
-  }
-
   final class PingRunnable extends NamedRunnable {
     final boolean reply;
     final int payload1;
     final int payload2;
-    final Ping ping;
 
-    PingRunnable(boolean reply, int payload1, int payload2, Ping ping) {
+    PingRunnable(boolean reply, int payload1, int payload2) {
       super("OkHttp %s ping %08x%08x", hostname, payload1, payload2);
       this.reply = reply;
       this.payload1 = payload1;
       this.payload2 = payload2;
-      this.ping = ping;
     }
 
     @Override public void execute() {
-      try {
-        // TODO(jwilson): fail the connection if replies are missing (and this is not a reply).
-        writePing(reply, payload1, payload2, ping);
-      } catch (IOException e) {
+      writePing(reply, payload1, payload2);
+    }
+  }
+
+  void writePing(boolean reply, int payload1, int payload2) {
+    if (!reply) {
+      int sentPingCount;
+      int receivedPongCount;
+      synchronized (this) {
+        sentPingCount = this.sentPingCount;
+        receivedPongCount = this.receivedPongCount;
+        this.sentPingCount++;
+      }
+      if (sentPingCount > receivedPongCount) {
```

Comment:
similar to the websocket case, do we ever see unrequested PING-ACKs from servers in the wild?

Reply:
No idea! Will address in follow up to track whether we're awaiting a pong rather than the counts.